File: frontend/lib/screens/monitoring/models/device_data.dart
import 'dart:convert';

// ฟังก์ชันแปลง JSON List เป็น List<DeviceData> (แม้ว่าตอนนี้เราจะใช้ decoder ใน ApiService แล้ว แต่เก็บไว้ก็ดี)
List<DeviceData> deviceDataListFromJson(String str) => 
    List<DeviceData>.from(json.decode(str).map((x) => DeviceData.fromJson(x)));

class DeviceData {
  final String? id; // _id
  final DateTime? timestamp;
  final String? name;
  final String? status;

  final double? lng;
  final double? lat;
  final int? rssi; 
  final double? snr;

  final double? acW;
  final double? acA;
  final double? acV;
  
  final int? battery;
  final int? lighting;
  final int? oat; 
  
  final Map<String, dynamic>? meta;

  DeviceData({
    this.id,
    this.timestamp,
    this.name,
    this.status,
    this.lng,
    this.lat,
    this.rssi,
    this.snr,
    this.acW,
    this.acA,
    this.acV,
    this.battery,
    this.lighting,
    this.oat,
    this.meta,
  });

  factory DeviceData.fromJson(Map<String, dynamic> json) {
    // Helper function สำหรับการดึงค่าจาก MongoDB format 
    String? extractMongoId(dynamic data) {
      if (data is Map && data.containsKey('\$oid')) {
        return data['\$oid'] as String?;
      }
      return null;
    }

    DateTime? extractMongoDate(dynamic data) {
      if (data is Map && data.containsKey('\$date')) {
        return DateTime.tryParse(data['\$date'] as String? ?? '');
      }
      if (data is String) {
        return DateTime.tryParse(data);
      }
      return null;
    }

    return DeviceData(
      id: extractMongoId(json['_id']) ?? json['_id'] as String?,
      timestamp: extractMongoDate(json['timestamp']),
      name: json['name'] as String?,
      status: json['status'] as String?,

      lng: (json['lng'] as num?)?.toDouble(),
      lat: (json['lat'] as num?)?.toDouble(),
      rssi: json['rssi'] as int?,
      snr: (json['snr'] as num?)?.toDouble(),
      
      acW: (json['acW'] as num?)?.toDouble(),
      acA: (json['acA'] as num?)?.toDouble(),
      acV: (json['acV'] as num?)?.toDouble(),

      battery: json['battery'] as int?,
      lighting: json['lighting'] as int?,
      oat: json['oat'] as int?,

      meta: json['meta'] as Map<String, dynamic>?,
    );
  }
}

File: backend/models/DeviceData.js
// // D:\mass_smart_city\Smart-Control\backend\models\DeviceData.js

// const mongoose = require('mongoose');

// // กำหนด Schema สำหรับ Collection deviceData
// // เราจะระบุฟิลด์ที่สำคัญเพื่อให้ Mongoose รู้จัก
// const deviceDataSchema = new mongoose.Schema(
//     {
//         // timestamp: { type: Date, required: true },
//         // meta: { type: Object }, // เก็บข้อมูล Object ภายใน เช่น name
//         acW: { type: Number }, // กำลังไฟ AC Watt
//         no: {type: Number},
//         // status: { type: String }, // สถานะ on/off
//         oat: { type: Number }, // อุณหภูมิภายนอก (Outdoor Ambient Temperature)
//         // lighting: { type: Number },
//         // battery: { type: Number },
//         // snr: { type: Number }, // Signal-to-Noise Ratio
//         acV: { type: Number }, // แรงดันไฟ AC Volt
//         lat: { type: Number }, // Latitude
//         lng: { type: Number }, // Longitude
//         acA: { type: Number }, // กระแสไฟ AC Ampere
//         // rssi: { type: Number }, // Received Signal Strength Indicator
//         // Mongoose จะจัดการ _id ให้เอง
//     },
//     { timestamps: true, collection: 'deviceData' } // ระบุชื่อ Collection ให้ตรงกับ MongoDB
// );

// module.exports = mongoose.model('DeviceData', deviceDataSchema);
const mongoose = require('mongoose');

// กำหนด Schema สำหรับ Collection deviceData (แบบ simplified ตาม requirement ล่าสุด)
const deviceDataSchema = new mongoose.Schema(
  {
    timestamp: { type: Date, required: true },

    // meta ใช้เก็บข้อมูลประกอบ เช่น no, deviceId ฯลฯ
    meta: { type: Object, default: {} },

    vac: { type: Number }, // แรงดัน DC Volt
    wac: { type: Number }, // กำลังไฟ DC Watt
    iac: { type: Number }, // กระแส DC Ampere
    acfreq: { type: Number }, // ความถี่ AC Hz
    acenergy: { type: Number }, // พลังงาน AC kWh
    vdc: { type: Number },
    idc: { type: Number },
    wdc: { type: Number },
    flag: { type: String }, // เก็บค่า flag ดิบ เช่น "$00" หรือ "$00"
    oat: { type: Number },
    lat: { type: Number },
    lng: { type: Number },
  },
  {
    collection: 'deviceData',
    timestamps: false,
    versionKey: false,
  }
);

// index สำหรับ query ล่าสุด / กราฟ
deviceDataSchema.index({ timestamp: -1 });

module.exports = mongoose.model('DeviceData', deviceDataSchema);
